app/models/user.rb

```diff
@@ -23,6 +24,17 @@ def self.find_or_create_from_auth_hash! auth_hash
     # end
   end
 
+  # Similar to has_secure_password's #authenticate, this tries to do a time
+  # constant comparison. Digesting both tokens before hand ensures that both
+  # strings are the same length during comparison. Suppose something could be
+  # done to store the digest instead, but that'd require overriding
+  # .has_secure_token
+  def token_authenticate token
+    token_digest = ::Digest::SHA256.hexdigest token
+    user_digest  = ::Digest::SHA256.hexdigest auth_token
```

Comment:
Some people go a little farther than this simply because a compromised database means compromised api tokens. One way around that is something like this:

```ruby
def generate_token
  pepper = ENV['API_TOKEN_PEPPER']
  @stored_token = SecureRandom.uuid
  @token_sent_to_user = Digest::SHA256.hexdigest "#{@stored_token}#{pepper}"
end

def token_authenticate token_sent_by_user
  pepper = ENV['API_TOKEN_PEPPER']
  valid_token = Digest::SHA256.hexdigest "#{@stored_token}#{pepper}"
  ActiveSupport::SecurityUtils.secure_compare valid_token, token_sent_by_user
end
```

or something like that

Reply:
Hrm, yeah that's fair and probably wise to do